package uk.gov.justice.digital.hmpps.communitypaybackapi.unit.service

import io.mockk.every
import io.mockk.impl.annotations.InjectMockKs
import io.mockk.impl.annotations.RelaxedMockK
import io.mockk.junit5.MockKExtension
import io.mockk.verify
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import uk.gov.justice.digital.hmpps.communitypaybackapi.client.CommunityPaybackAndDeliusClient
import uk.gov.justice.digital.hmpps.communitypaybackapi.common.validation.ValidationResult
import uk.gov.justice.digital.hmpps.communitypaybackapi.dto.AppointmentDto
import uk.gov.justice.digital.hmpps.communitypaybackapi.dto.ProjectDto
import uk.gov.justice.digital.hmpps.communitypaybackapi.dto.UpdateAppointmentDto
import uk.gov.justice.digital.hmpps.communitypaybackapi.dto.exceptions.ConflictException
import uk.gov.justice.digital.hmpps.communitypaybackapi.dto.exceptions.InternalServerErrorException
import uk.gov.justice.digital.hmpps.communitypaybackapi.factory.dto.valid
import uk.gov.justice.digital.hmpps.communitypaybackapi.factory.entity.valid
import uk.gov.justice.digital.hmpps.communitypaybackapi.service.AppointmentEventService
import uk.gov.justice.digital.hmpps.communitypaybackapi.service.AppointmentEventTrigger
import uk.gov.justice.digital.hmpps.communitypaybackapi.service.AppointmentRetrievalService
import uk.gov.justice.digital.hmpps.communitypaybackapi.service.AppointmentUpdateService
import uk.gov.justice.digital.hmpps.communitypaybackapi.service.AppointmentValidationService
import uk.gov.justice.digital.hmpps.communitypaybackapi.service.UpdateAppointmentValidationService
import uk.gov.justice.digital.hmpps.communitypaybackapi.service.internal.SpringEventPublisher
import uk.gov.justice.digital.hmpps.communitypaybackapi.service.mappers.ToAppointmentEntity.toAppointmentEntity
import uk.gov.justice.digital.hmpps.communitypaybackapi.unit.util.WebClientResponseExceptionFactory

@ExtendWith(MockKExtension::class)
class AppointmentUpdateServiceTest {

  @RelaxedMockK
  lateinit var appointmentRetrievalService: AppointmentRetrievalService

  @RelaxedMockK
  lateinit var appointmentEventService: AppointmentEventService

  @RelaxedMockK
  lateinit var communityPaybackAndDeliusClient: CommunityPaybackAndDeliusClient

  @RelaxedMockK
  lateinit var updateAppointmentValidationService: UpdateAppointmentValidationService

  @RelaxedMockK
  lateinit var springEventPublisher: SpringEventPublisher

  @InjectMockKs
  lateinit var service: AppointmentUpdateService

  companion object {
    const val PROJECT_CODE = "PROJ123"
    const val DELIUS_APPOINTMENT_ID = 101L
    val TRIGGER = AppointmentEventTrigger.valid()
  }

  @Nested
  inner class UpdateAppointmentOutcome {

    val existingAppointment = AppointmentDto.valid().copy(id = DELIUS_APPOINTMENT_ID, projectCode = PROJECT_CODE)
    val updateRequest = UpdateAppointmentDto.valid().copy(deliusId = DELIUS_APPOINTMENT_ID)

    @Test
    fun `if there's no existing entries for the delius appointment ids, persist new entry, raise domain event and invoke update endpoint`() {
      val appointmentEntity = existingAppointment.toAppointmentEntity(null, null, null)
      every { appointmentRetrievalService.getOrCreateAppointmentEntity(existingAppointment) } returns appointmentEntity
      every { updateAppointmentValidationService.validate(any(), any()) } answers {
        val ctx = it.invocation.args[1] as AppointmentValidationService.AppointmentValidationContext.Update
        ctx.project = ProjectDto.valid()
        ValidationResult.success()
      }

      service.updateAppointment(
        existingAppointment = existingAppointment,
        update = updateRequest,
        trigger = TRIGGER,
      )

      verify {
        springEventPublisher.publishEvent(any())
        communityPaybackAndDeliusClient.updateAppointment(
          projectCode = PROJECT_CODE,
          appointmentId = DELIUS_APPOINTMENT_ID,
          updateAppointment = any(),
        )
      }
    }

    @Test
    fun `if there's an existing entry for the delius appointment id and it's logically identical, do not send an update`() {
      every { updateAppointmentValidationService.validate(any(), any()) } answers {
        val ctx = it.invocation.args[1] as AppointmentValidationService.AppointmentValidationContext.Update
        ctx.project = ProjectDto.valid()
        ValidationResult.success()
      }
      every { appointmentEventService.hasUpdateAlreadyBeenSent(any()) } returns true

      service.updateAppointment(
        existingAppointment = existingAppointment,
        update = updateRequest,
        trigger = TRIGGER,
      )

      verify(exactly = 0) {
        springEventPublisher.publishEvent(any())
        communityPaybackAndDeliusClient.updateAppointment(any(), any(), any())
      }
    }

    @Test
    fun `if appointment has newer version on update, throw conflict exception`() {
      every { updateAppointmentValidationService.validate(any(), any()) } answers {
        val ctx = it.invocation.args[1] as AppointmentValidationService.AppointmentValidationContext.Update
        ctx.project = ProjectDto.valid()
        ValidationResult.success()
      }

      every { appointmentEventService.hasUpdateAlreadyBeenSent(any()) } returns false

      every {
        communityPaybackAndDeliusClient.updateAppointment(any(), any(), any())
      } throws WebClientResponseExceptionFactory.conflict()

      assertThatThrownBy {
        service.updateAppointment(
          existingAppointment = existingAppointment,
          update = updateRequest,
          trigger = TRIGGER,
        )
      }.isInstanceOf(ConflictException::class.java).hasMessage("A newer version of the appointment exists. Stale version is '${updateRequest.deliusVersionToUpdate}'")
    }

    @Test
    fun `if bad request returned throw internal server error`() {
      every { updateAppointmentValidationService.validate(any(), any()) } answers {
        val ctx = it.invocation.args[1] as AppointmentValidationService.AppointmentValidationContext.Update
        ctx.project = ProjectDto.valid()
        ValidationResult.success()
      }
      every { appointmentEventService.hasUpdateAlreadyBeenSent(any()) } returns false

      every {
        communityPaybackAndDeliusClient.updateAppointment(any(), any(), any())
      } throws WebClientResponseExceptionFactory.badRequest("didn't look good")

      assertThatThrownBy {
        service.updateAppointment(
          existingAppointment = existingAppointment,
          update = updateRequest,
          trigger = TRIGGER,
        )
      }.isInstanceOf(InternalServerErrorException::class.java).hasMessage("Bad request returned updating an appointment. Upstream response is 'didn't look good'")
    }
  }
}
